### PR TITLE
one allocate per resource per pod

### LINKED ISSUE #66804: Issues a single Allocate gRPC call per pod in kubelet device manager
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind feature

**What happened**:

**What you expected to happen**:
Basically implements this TODO in device manager allocateContainerResources():
"TODO: refactor this part of code to just append a ContainerAllocationRequest
in a passed in AllocateRequest pointer, and issues a single Allocate call per pod."

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by tianshapjq**
**What this PR does / why we need it**:
this is my first draft for issuing the "one allocate per pod", still have something to figure out:
1. I add deviceId in the DeviceSpec to locate the specific device passed to DP, and update back after issuing the request to DP.
2. This approach assumes that the Envs, Mounts and Annotations are all equal for all devices of the same resource, is this feasible?
3. this would make the interface version to v1beta2, which is not compatible with the v1beta1. This is what i concern most about, do we need to make the interface compatible with v1beta1?
4. maybe passing AllocateRequest, which could contain the mapping of containerName and request devices, to DP directly would be better? But this is still not compatible with the v1beta1, because the current Allocate() interface just accepts []string.

This is my first draft, I am not sure if adding deviceId to DeviceSpec is feasible, would like to hear your feedbacks!

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #66804

**Special notes for your reviewer**:
@jiayingz @mtaufen @liggitt 

**Release note**:

```release-note
Optimize the allocate calls to device plugin, issuing one allocate call to each resource per pod, would reduce the number of allocate calls.
```
/sig node
/area kubelet
/assign @jiayingz

### CODE REVIEW

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go [hunk 1]
```diff
@@ -1134,6 +1143,12 @@ func (m *DeviceSpec) MarshalTo(dAtA []byte) (int, error) {
 		i = encodeVarintApi(dAtA, i, uint64(len(m.Permissions)))
 		i += copy(dAtA[i:], m.Permissions)
 	}
+	if len(m.DeviceId) > 0 {
+		dAtA[i] = 0x22
+		i++
+		i = encodeVarintApi(dAtA, i, uint64(len(m.DeviceId)))
+		i += copy(dAtA[i:], m.DeviceId)
+	}
```
Comment (neolit123) [inline pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: is this covered well by unit tests?

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go [hunk 2]
```diff
@@ -2967,6 +2987,35 @@ func (m *DeviceSpec) Unmarshal(dAtA []byte) error {
...
+			intStringLen := int(stringLen)
+			if intStringLen < 0 {
+				return ErrInvalidLengthApi
+			}
+			postIndex := iNdEx + intStringLen
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.DeviceId = string(dAtA[iNdEx:postIndex])
+			iNdEx = postIndex
```
Comment (neolit123) [inline pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: looks like this function might need some helper functions eventually.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 3]
```diff
@@ -271,19 +276,63 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
 	devicesToReuse := make(map[string]sets.String)
+	// key by resource name, representing all the devices to be allocated of one resource
+	resourceDevices := make(map[string]sets.String)
+
+	// step1. aggregate all devices requested by containers
 	for _, container := range pod.Spec.InitContainers {
-		if err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
+		if devices, err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
 			return err
+		} else {
```
Comment (neolit123) [inline pkg/kubelet/cm/devicemanager/manager.go]: please replace:
```
} else {
	// union devices to the underlying resource
	unionDevices(resourceDevices, devices)
}
```

with:
```
}
// union devices to the underlying resource
unionDevices(resourceDevices, devices)
```

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 4]
```diff
@@ -271,19 +276,63 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
+			// union devices to the underlying resource
+			unionDevices(resourceDevices, devices)
 		}
 		m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, devicesToReuse)
 	}
 	for _, container := range pod.Spec.Containers {
-		if err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
+		if devices, err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
 			return err
+		} else {
```
Comment (neolit123) [inline pkg/kubelet/cm/devicemanager/manager.go]: no need for `} else {`

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 5]
```diff
@@ -271,19 +276,63 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
+			unionDevices(resourceDevices, devices)
 		}
 		m.podDevices.removeContainerAllocatedResources(string(pod.UID), container.Name, devicesToReuse)
 	}
 
+	// step 2. issue rpc request for each resource and update back to the podDevices
+	for resource := range resourceDevices {
+		if response, err := m.allocateResources(resource, resourceDevices[resource]); err != nil {
+			return err
+		} else {
```
Comment (neolit123) [inline pkg/kubelet/cm/devicemanager/manager.go]: no need for `} else {`

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 6]
```diff
@@ -608,14 +667,66 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
...
+	// plugin Allocate grpc calls if it becomes common that a container may require
+	// resources from multiple device plugins.
+	m.mutex.Lock()
+	e, ok := m.endpoints[resource]
+	m.mutex.Unlock()
+	if !ok {
+		m.mutex.Lock()
+		m.allocatedDevices = m.podDevices.devices()
+		m.mutex.Unlock()
+		return nil, fmt.Errorf("Unknown Device Plugin %s", resource)
```
Comment (neolit123) [inline pkg/kubelet/cm/devicemanager/manager.go]: errors should start with lowercase.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 7]
```diff
@@ -608,14 +667,66 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
...
+		// In case of allocation failure, we want to restore m.allocatedDevices
+		// to the actual allocated state from m.podDevices.
+		m.mutex.Lock()
+		m.allocatedDevices = m.podDevices.devices()
+		m.mutex.Unlock()
+		return nil, err
+	}
+
+	if len(resp.ContainerResponses) == 0 {
+		return nil, fmt.Errorf("No containers return in allocation response %v", resp)
```
Comment (neolit123) [inline pkg/kubelet/cm/devicemanager/manager.go]: errors should start with lowercase.

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go (hunk 1 above)
Comment (tianshapjq) [reply pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: no, this is auto-generated

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 3 above)
Comment (tianshapjq) [reply pkg/kubelet/cm/devicemanager/manager.go]: I am still not sure how feasible this modification could be. I will make more optimizations after other reviews :)

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go [hunk 8]
```diff
@@ -361,6 +361,8 @@ type DeviceSpec struct {
 	// * w - allows container to write to the specified device.
 	// * m - allows container to create device files that do not yet exist.
 	Permissions string `protobuf:"bytes,3,opt,name=permissions,proto3" json:"permissions,omitempty"`
+	// id to identify the specific device
+	DeviceId string `protobuf:"bytes,4,opt,name=device_id,json=deviceId,proto3" json:"device_id,omitempty"`
```
Comment (liggitt) [inline pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: field name should be `DeviceID` with json of `deviceID`, right?

**Comment by liggitt** [comment]
one nit on the API field name, I did not review the actual logic using the field

**Comment by liggitt** [comment]
@kubernetes/sig-node-api-reviews

**Comment by jiayingz** [comment]
Thanks a lot for the change! I do have concern on the API change. Could you explain why that is necessary? I think device manager knows what devices to allocate for each container before issuing the gRPC call and we request that a device plugin'ss AllocateResponse has ContainerAllocateResponse in the same order as AllocateRequest.ContainerAllocateRequest. With that, can we get allocated device ids from device manager's local data structure instead of making the API change?

**Comment by vikaschoudhary16** [comment]
@tianshapjq Though i have few concerns on the code changes but before commenting on code changes, lets first discuss what this PR's objective is.
It seems you want to merge per container gRPC call to a single call. I would have liked this idea if no DPAPI changes were required for this. To break a beta api seems not a good idea for this optimization IMO. Do you have any numbers to back this optimization?

**Comment by tianshapjq** [comment]
@vikaschoudhary16 yeah, I agree that breaking a beta api is quite rude to users. Sorry I don't get clear with your number, what does it mean, such as v1beta2?
Without breaking the current API, just as what I describe above, we can store the mapping of device IDs and their requested order in local in case that devices in DP response keeps the exact order with request. This would stabilize the current API. Right now I just have to make sure that the response has the same order with request, because we don't explicitly ask users to do so at current.

**Comment by jiayingz** [comment]
+1 to store the mapping of device IDs and their requested order in local cache and document explicitly in pkg/kubelet/apis/deviceplugin/v1beta1/api.proto that device plugins should keep the same order in the Allocate response.

**Comment by RenaudWasTaken** [comment]
/ok-to-test

I didn't read the code in itself but if this is about changing the n-Allocate calls into 1 global Allocate call then we can do that by still supporting both versions.

Add an attribute to the endpoint that identifies if it is a v1beta1 or v1beta2 and based on the version issue n-Allocate calls or 1-Allocate call.

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go (hunk 2 above)
Comment (RenaudWasTaken) [reply pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: this is auto-generated, we shouldn't make any changes to it

**Comment by tianshapjq** [comment]
@RenaudWasTaken Thanks for your advice! I would prefer storing mapping of device IDs and requested order in local instead of newing a v1beta2 version, this is enough for the "one allocate per resource" at present. So it would be better to keep the current API stable IMO. WDYT?

**Comment by RenaudWasTaken** [comment]
> Add an attribute to the endpoint that identifies if it is a v1beta1 or v1beta2 and based on the version issue n-Allocate calls or 1-Allocate call.

Actually it's not the only thing, you're going to hit type issue, you'll need to create a new implementation of the endpoint interface that uses that type.

Maybe we could move the logic of the n-Allocator vs 1-Allocate into the endpoint?

**Comment by tianshapjq** [comment]
> Actually it's not the only thing, you're going to hit type issue, you'll need to create a new implementation of the endpoint interface that uses that type.
Yes, I agree that we will hit that issue sooner or later, such as adding a new type or changing the current. Then we have to split several pieces of logic to deal with v1beta1, v1beta2 etc. That would be another job for optimization, right now I would like to focus on the 1-Allocate ability and be sure to make least changes to the beta API.

> Maybe we could move the logic of the n-Allocator vs 1-Allocate into the endpoint?
maybe moving this logic into endpoint would not a better choice, endpoint would keep as clearly as possible to deal with DPs. In addition to passing podDevices to the endpoint, we also need to move a lot of other logic into the endpoint, which increases the complexity of the endpoint. I'd rather choosing to abstract it somewhere instead of moving into endpoint.

**Comment by tianshapjq** [comment]
> Add an attribute to the endpoint that identifies if it is a v1beta1 or v1beta2 and based on the version issue n-Allocate calls or 1-Allocate call.

there is already an attribute named Version in RegisterRequest to identify the interface inversion supported by DP :)

**Comment by tianshapjq** [comment]
@jiayingz I found without passing deviceId may not be possible, cause the DP may return more devices other than the requested ones. Such as:
require gpu nvidia1 then the DP would return "nvidia1", "nvidiactl" and "nvidia-uvm", this would break the rule of keeping exact order with request. Or maybe just the first half of the reponse must keep order with request, and the rest will all mount into container? This is a somewhat ugly and too customized.

**Comment by tianshapjq** [comment]
maybe v1beta2 is needed

**Comment by jiayingz** [comment]
@tianshapjq the devices in DeviceSpec you mentioned is different from the deviceIds we generally talk about in device plugin. The DeviceSpec is part of runtime configuration and can be multiple. On the other hand, the deviceIds we get from DP ListWatch is the resource allocation units.

I don't think we need a new API version.

**Comment by tianshapjq** [comment]
@jiayingz I get that device IDs from DP ListAndWatch is the units for allocation, but we also require IDs in the DeviceSpec. According disscussion with @vikaschoudhary16, Considering this:

container1 requested gpuid-100 and 101 of resource nvidia.com/gpu
container2 requested gpuid-200 and 201 of resource nvidia.com/gpu

I combine these 2 requests together, and send them to endpoint only one time, that would be like:
pod
    resource[nvidia.com/gpu] = gpu100 gpu101 gpu200 gpu201

and then the current response would like:
    envs ["aaa","bbb"]
    Mounts ["/dev":"/dev"]
    Devices [{containerPath:"/dev/gpu100", hostPath:"/dev/gpu100"}, {containerPath:"/dev/gpu101", hostPath:"/dev/gpu101"}, {containerPath:"/dev/gpu200", hostPath:"/dev/gpu200"}, {containerPath:"/dev/gpu201", hostPath:"/dev/gpu201"}, {containerPath:"/dev/nvidiactl", hostPath:"/dev/nvidiactl"}, {containerPath:"/dev/nvidia-uvm", hostPath:"/dev/nvidia-uvm"}]
    Annotations ...

notice that nvidiactl and nvidia-uvm doesn't contain in the requests, then without IDs in Devices how can I tell the {"/dev/gpu201", "/dev/gpu201"}, {"/dev/nvidiactl","/dev/nvidiactl"}, {"/dev/nvidia-uvm","/dev/nvidia-uvm"}} is belonging to container2?

So what I want in this PR is changing the response like this:
    Devices [{id:"gpu100", containerPath:"/dev/gpu100", hostPath:"/dev/gpu100"}, {id:"gpu101", containerPath:"/dev/gpu101", hostPath:"/dev/gpu101"}, {id:"/dev/gpu200", hostPath:"/dev/gpu200"}, {id:"/dev/gpu201", hostPath:"/dev/gpu201"}, {id:"", hostPath:"/dev/nvidiactl"}, {id:"", hostPath:"/dev/nvidia-uvm"}]

the devices with id=="" will all mount into container requesting this resource

**Comment by jiayingz** [comment]
ContainerAllocateRequest is a repeated field in AllocateRequest, and so is ContainerAllocateResponse in AllocateResponse.

In the above example, the AllocateRequest would be:
```
AllocateRequest {
  ContainerAllocateRequest {
      DeviceIDs: [gpuid-100", "gpuid-101"]
  }
  ContainerAllocateRequest {
      DeviceIDs: [gpuid-102", "gpuid-103"]
  }
}
```
The AllocateResponse is:
```
AllocateResponse {
  ContainerAllocateResponse {
       Envs: ["aaa","bbb"]
       Mounts: ["/dev":"/dev"]
       Devices: [{containerPath:"/dev/gpu100", hostPath:"/dev/gpu100"},
                       {containerPath:"/dev/gpu101", hostPath:"/dev/gpu101"},
                       {containerPath:"/dev/nvidiactl", hostPath:"/dev/nvidiactl"},
                       {containerPath:"/dev/nvidia-uvm", hostPath:"/dev/nvidia-uvm"}]
   }
  ContainerAllocateResponse {
       Envs: ["aaa","bbb"]
       Mounts: ["/dev":"/dev"]
       Devices: [{containerPath:"/dev/gpu102", hostPath:"/dev/gpu102"},
                       {containerPath:"/dev/gpu103", hostPath:"/dev/gpu103"},
                       {containerPath:"/dev/nvidiactl", hostPath:"/dev/nvidiactl"},
                       {containerPath:"/dev/nvidia-uvm", hostPath:"/dev/nvidia-uvm"}]
   }
}
```

File: pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go (hunk 8 above)
Comment (vishh) [reply pkg/kubelet/apis/deviceplugin/v1beta1/api.pb.go]: +1

**Comment by vishh** [comment]
Being able to associate individual allocate response to specific allocate requests in a batch allocate request makes sense to me at an API level. Adding a new field shouldn't require a new API version. 
Requiring strict ordering is a subtle requirement being imposed on device plugin authors and many of them may overlook it.
For the sake of backwards compatibility, it will still be necessary to include the client side caching/lookup logic mentioned in PR feedback.
Thoughts?

**Comment by tianshapjq** [comment]
@jiayingz thanks for your feedback! What you describe is what we are doing right now, but my changes are going to combine that two ContainerAllocateRequests together and send them to DP in one allocate, because these two ContainerAllocateRequests are requring for the same resource.
So, the AllocateRequest would be:
```
AllocateRequest {
  ContainerAllocateRequest {
      DeviceIDs: ["gpuid-100", "gpuid-101", "gpuid-102", "gpuid-103"]
  }
}
```
and the AllocateResponse would be:
```
AllocateResponse {
  ContainerAllocateResponse {
       Envs: ["aaa","bbb"]
       Mounts: ["/dev":"/dev"]
       Devices: [{containerPath:"/dev/gpu100", hostPath:"/dev/gpu100"},
                       {containerPath:"/dev/gpu101", hostPath:"/dev/gpu101"},
                       {containerPath:"/dev/gpu102", hostPath:"/dev/gpu102"},
                       {containerPath:"/dev/gpu103", hostPath:"/dev/gpu103"},
                       {containerPath:"/dev/nvidiactl", hostPath:"/dev/nvidiactl"},
                       {containerPath:"/dev/nvidia-uvm", hostPath:"/dev/nvidia-uvm"}]
   }
}
```

**Comment by jiayingz** [comment]
@tianshapjq we don't want to combine ContainerAllocateRequests together. That is going to cause wrong behavior in some cases. Each ContainerAllocateRequest maps to a container in the pod that requests that resource.

**Comment by tianshapjq** [comment]
@vishh I do have this concern for DP authors that they would not notice about the strict order. But without introducing a new API version, we can't tell if DP supports contianing deviceId in the response, which would not be able to associate the device of response to the specific container.

**Comment by tianshapjq** [comment]
@jiayingz sorry I don't get your point clearly, I thought the original issue is about to call one allocate to one resource at a time, not separately by each container, did I get it wrong?

> Each ContainerAllocateRequest maps to a container in the pod that requests that resource.

this is what we do at current, we may not change anything if I get it wrong.

**Comment by vishh** [comment]
> we can't tell if DP supports contianing deviceId in the response

Can you elaborate on this statement a bit more? Which scenario are you considering - old client vs new server or new client vs old server? 
How do we plan on adding fields after the API hits GA (v1)?

**Comment by tianshapjq** [comment]
> Which scenario are you considering - old client vs new server or new client vs old server?

it's the old client vs new server. What I'm trying to do is to combine devices belonging to same resource in one pod and send them to DP only once, while the present codes would send them multiple times by each container. 
In my changes, if client doesn't respond devices with id, it won't be able to map the response device to specific container, cause my changes combine the requested devices and allocate DP only once.
On the other hand, if adding deviceId in the DeviceSpec, we have to introduce a new v1beta2 to let server know that the client will respond with id, and server will do the combination job, otherwise the server will do allocations in the v1beta1, that is, not combine devices together and keep one allocate per container.

> How do we plan on adding fields after the API hits GA (v1)?

that's what I concern about, we may have to support both v1beta1 and v1beta2 at the same time, that would be as what @RenaudWasTaken said:

> Add an attribute to the endpoint that identifies if it is a v1beta1 or v1beta2 and based on the version issue n-Allocate calls or 1-Allocate call.

**Comment by vishh** [comment]
Just to be on the same page, in my mind, the kubelet is the client and device plugins are the servers. 

> otherwise the server will do allocations in the v1beta1

If for example you add a scalar string field, or a new message to detect the field being unset, on the client side you can check if the field is unset and default to the in-memory map you plan on maintaining. If it is set, you can ditch the in-memory map. 

>  we may have to support both v1beta1 and v1beta2 

I don't think this answers my question. What do we do if we add fields after v1 - do we introduce a new go package every time the API is extended?

**Comment by vikaschoudhary16** [comment]
@tianshapjq 
> @jiayingz sorry I don't get your point clearly, I thought the original issue is about to call one allocate to one resource at a time, not separately by each container, did I get it wrong?

I think @jiayingz was suggesting that AllocateRequest is a list of ContainerAllocateRequest:
```
type AllocateRequest struct {
        ContainerRequests []*ContainerAllocateRequest `protobuf:"bytes,1,rep,name=container_requests,json=containerRequests" json:"container_requests,omitempty"`
}
```
So, it seems possible to make one AllocateRequest per pod per resource where each element in the list i.e ContainerAllocateRequest is for a specific container in the pod. After creating AllocateRequest like this, only one gRPC call for one resource will work with the limitation that order of elements in the ContainerAllocateResponses from the plugin must be in the same order as the order of ContainerAllocateRequests in the AllocateRequest.

However, agree that plugin authors may miss this point. 

@vishh 
> I don't think this answers my question. What do we do if we add fields after v1 - do we introduce a new go package every time the API is extended?

In grpc, a new field or new method is not a breaking change. New client should handle backward compatibility by either handling "unimplemented error", or introduce some kind of feature discovery for example through device plugin options. It seems needless to introduce a new go package every time the API is extended.

@vishh @jiayingz @RenaudWasTaken, how about this approach:
Add a new option, `BatchedAllocateRequestSupport` in `DevicePluginOptions` and `deviceID` in DeviceSpec. By default it will be false and device manager will work as it works today to support old plugins. `deviceID` will simple be ignored and not used in device manager.
New plugins, with the batched feature support, will send this flag as `true`.
If `BatchedAllocateRequestSupport` is received true, device manager will execute the logic similar to the changes in this PR and will expect initialized `deviceID`, otherwise it will error-out.
I dont think we need to bump api version for the above approach.

**Comment by tianshapjq** [comment]
@vikaschoudhary16 thanks for your explanation! Yes, with the new type AllocateRequest it's possible to make only one allocate per resource. But we also have to introduce a new interface, say allocateBatch(), besides the current allocate(devs []string)  in endpoint, right?

**Comment by vikaschoudhary16** [comment]
> But we also have to introduce a new interface, say allocateBatch(), besides the current allocate(devs []string) in endpoint, right?

That is not going to affect DPAPI and plugins, so should not be an issue.

**Comment by jiayingz** [comment]
This seems add too much complexity on device mamager side. I also feel list of device IDs perhaps is not good identifier for a container allocation request. To limit the scope of this PR, can we just implement the merging of gRPC call? Or if people feel the ordering requirement is subtle, maybe we should hold the change for now till we have settled down the device plugin life cycle management.

**Comment by vikaschoudhary16** [comment]
Agree adding deviceid and managing without version bump going to increase somewhat complexity on the manager side.  
> ordering requirement is subtle, maybe we should hold the change for now till we have settled down the device plugin life cycle management.

+1

**Comment by tianshapjq** [comment]
> To limit the scope of this PR, can we just implement the merging of gRPC call?

@jiayingz sorry I didn't get your point, what's the merging of gRPC call? Do you mean adding ContainerAllocateRequest or deviceId in the interface at first?

**Pull request closed without merging**